polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp

```diff
@@ -535,35 +544,43 @@ Value castToIndex(Value val) {
 // WorkGroupSize implementation
 //===----------------------------------------------------------------------===//
 
-WorkGroupSize::WorkGroupSize(Value ndItem,
+WorkGroupSize::WorkGroupSize(unsigned numDims,
                              const sycl::ReqdWorkGroupSize &reqdWorkGroupSize,
                              OpBuilder builder) {
-  auto ndItemTy = cast<sycl::NdItemType>(
-      cast<MemRefType>(ndItem.getType()).getElementType());
-  const unsigned numDims = ndItemTy.getDimension();
   assert((reqdWorkGroupSize.empty() || reqdWorkGroupSize.size() == numDims) &&
          "Expecting reqdWorkGroupSize to have same number of elements as "
-         "nd_item dimension");
-  for (unsigned dim = 0; dim < numDims; ++dim) {
-    /// Use constants provided by reqd_work_group_size if given (non-empty).
-    if (!reqdWorkGroupSize.empty())
+         "numDims");
+
+  // Use constants provided by reqd_work_group_size if given (non-empty).
+  if (!reqdWorkGroupSize.empty()) {
+    for (unsigned dim = 0; dim < numDims; ++dim)
       wgSizes.push_back(reqdWorkGroupSize[dim]);
-    else
-      wgSizes.push_back(castToIndex(getLocalRange(ndItem, dim, builder)));
+    return;
   }
+
+  populateWorkGroupSize(numDims, builder);
 }
 
 WorkGroupSize::ElemTy WorkGroupSize::operator[](unsigned dim) const {
   assert(dim < wgSizes.size() && "Expecting valid dim");
   return wgSizes[dim];
 }
 
-Value WorkGroupSize::getLocalRange(Value ndItem, unsigned dim,
-                                   OpBuilder builder) const {
-  return builder.create<sycl::SYCLNDItemGetLocalRangeOp>(
-      builder.getUnknownLoc(), builder.getI64Type(), ndItem,
-      builder.create<arith::ConstantIntOp>(builder.getUnknownLoc(), dim,
-                                           builder.getI32Type()));
+void WorkGroupSize::populateWorkGroupSize(unsigned numDims, OpBuilder builder) {
+  const auto arrayType = builder.getType<sycl::ArrayType>(
+      numDims, MemRefType::get(numDims, builder.getIndexType()));
+  const auto rangeTy = builder.getType<sycl::RangeType>(numDims, arrayType);
+  Location loc = builder.getUnknownLoc();
+  auto wgSize = builder.create<sycl::SYCLWorkGroupSizeOp>(loc, rangeTy);
```

Comment:
Can you add a utility function `createWorkgroupSize` in the SYCL `Utils`  similar to what we already have (e.g. createSYCLRangeGetOp).

Reply:
done.